core/src/runtime/execution_report.rs

```diff
@@ -0,0 +1,171 @@
+/// };
+///
+/// let instructions = vec![
+///     Instruction::new(Opcode::ADD, 29, 0, 5, false, true),
+///     Instruction::new(Opcode::ADD, 30, 0, 37, false, true),
+///     Instruction::new(Opcode::ADD, 31, 30, 29, false, false),
+/// ];
+///
+/// // Setup a program and a runtime
+/// let program = Program::new(instructions, 0, 0);
+/// let mut runtime = Runtime::new(program, SP1CoreOpts::default());
+/// // Runtime must execute before reporting statistics
+/// let report = runtime.dry_run().unwrap();
+/// assert_eq!(runtime.register(Register::X31), 42);
+///
+/// // Retrieve the number of ADD operations
+/// let add_count = &report.get_metric("Arithmetic", "ADD").unwrap_or(0);
+/// assert_eq!(*add_count, 3);
+/// assert_eq!(report.cycles, 12);
+/// assert_eq!(report.total_instruction_count, 3);
+///
+/// // You can also get a count of all instruction/opcode occurrences by category
+/// let add_operations_count = report.get_total_for_category("Arithmetic").unwrap_or(0);
+/// assert_eq!(add_operations_count, 3,);
+/// ```
+#[derive(Debug)]
+pub struct ExecutionReport {
+    /// Total number of cycles occurred during program execution
+    pub cycles: u32,
+    /// Total number of instructions occurred during program execution
+    pub total_instruction_count: usize,
+    /// Total number of syscalls occurred during program execution
+    pub syscall_count: usize,
+    /// Contains precise counts of each individual instruction,
+    /// as well as counts of instruction by category. See [ExecutionReport] Usage for
+    /// more details
+    pub custom_metrics: HashMap<String, Box<dyn Any>>,
+}
+
+impl ExecutionReport {
+    /// Accept a runtime from a program which has already executed
+    /// and extract some useful information from it.
+    pub fn new(runtime: &Runtime) -> Self {
+        let total_instruction_count = runtime
+            .instruction_log
+            .values()
+            .map(|counts| counts.values().sum::<u64>() as usize)
+            .sum();
+
+        let syscall_count = runtime
+            .instruction_log
+            .get("Syscalls")
+            .map_or(0, |counts| counts.values().sum::<u64>() as usize);
+
+        let mut custom_metrics = HashMap::new();
+        for (category, counts) in &runtime.instruction_log {
+            // Convert the inner HashMap to a format that can be boxed as `dyn Any`
+            let metrics: HashMap<_, _> = counts
+                .iter()
+                .map(|(opcode, count)| (format!("{:?}", opcode), *count))
+                .collect();
+            custom_metrics.insert(category.clone(), Box::new(metrics) as Box<dyn Any>);
+        }
+
+        Self {
+            cycles: runtime.state.clk,
+            total_instruction_count,
+            syscall_count,
+            custom_metrics,
+        }
+    }
+
+    /// Retrieve specific metric based on category and opcode.
+    pub fn get_metric(&self, category: &str, opcode: &str) -> Result<u64, MetricRetrievalError> {
```

Comment:
these methods are unecessary, can just let users directly access fields on struct for now

Reply:
don't want to ensrhine categories inside of Runtime